File: src/data/codes/nonlinear-svm-rbf.m
clc;
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'rbf_svm_tumor.csv';

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(12);
    n = 160;

    Radius = 10 + 8*rand(n,1);
    Texture = 12 + 12*rand(n,1);
    Smoothness = 0.07 + 0.12*rand(n,1);
    Symmetry = 0.10 + 0.18*rand(n,1);

    % Nonlinear-like relation
    risk = (Radius - 14).^2 + (Texture - 18).^2 + 60*Smoothness + 40*Symmetry;
    Diagnosis = double(risk > median(risk));

    data = table(Radius, Texture, Smoothness, Symmetry, Diagnosis);
else
    data = readtable(csvFile);
    % Expected columns:
    % Radius, Texture, Smoothness, Symmetry, Diagnosis
end

disp(head(data));

%% =========================
% PREPROCESSING
% ==========================
data = rmmissing(data);
data = unique(data);

X = [data.Radius, data.Texture, data.Smoothness, data.Symmetry];
Y = data.Diagnosis;

X = normalize(X);

%% =========================
% TRAIN-TEST SPLIT
% ==========================
cv = cvpartition(Y, 'HoldOut', 0.2);
X_train = X(training(cv), :);
Y_train = Y(training(cv), :);
X_test  = X(test(cv), :);
Y_test  = Y(test(cv), :);

%% =========================
% MODEL BUILDING
% ==========================
model = fitcsvm(X_train, Y_train, ...
    'KernelFunction', 'rbf', ...
    'Standardize', true);

%% =========================
% PREDICTION
% ==========================
Y_pred = predict(model, X_test);

%% =========================
% EVALUATION
% ==========================
confMat = confusionmat(Y_test, Y_pred);
accuracy = sum(diag(confMat)) / sum(confMat(:));

TP = confMat(2,2); FP = confMat(1,2); FN = confMat(2,1);
precision = TP / (TP + FP + eps);
recall    = TP / (TP + FN + eps);
f1        = 2 * precision * recall / (precision + recall + eps);

disp('Confusion Matrix:');
disp(confMat);
fprintf('\nAccuracy  = %.4f\n', accuracy);
fprintf('Precision = %.4f\n', precision);
fprintf('Recall    = %.4f\n', recall);
fprintf('F1-score  = %.4f\n', f1);

figure;
confusionchart(Y_test, Y_pred);
title('RBF SVM Confusion Matrix');

File: src/data/codes/nonlinear-svm-rbf.py
"""SVM with RBF kernel
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

USE_SYNTHETIC = True
RNG = np.random.default_rng(12)

if USE_SYNTHETIC:
    n = 160
    radius = RNG.uniform(10, 18, size=n)
    texture = RNG.uniform(12, 24, size=n)
    smooth = RNG.uniform(0.07, 0.19, size=n)
    symm = RNG.uniform(0.10, 0.28, size=n)
    risk = (radius - 14) ** 2 + (texture - 18) ** 2 + 60 * smooth + 40 * symm
    med = np.median(risk)
    diagnosis = (risk > med).astype(int)
    data = pd.DataFrame(
        {
            "Radius": radius,
            "Texture": texture,
            "Smoothness": smooth,
            "Symmetry": symm,
            "Diagnosis": diagnosis,
        }
    )
else:
    data = pd.read_csv("rbf_svm_tumor.csv")

print(data.head())
data = data.dropna().drop_duplicates()
X = data[["Radius", "Texture", "Smoothness", "Symmetry"]].values
y = data["Diagnosis"].values
scaler = StandardScaler()
X = scaler.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

model = SVC(kernel="rbf", random_state=42).fit(X_train, y_train)
y_pred = model.predict(X_test)

cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", cm)
print(f"Accuracy  = {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision = {precision_score(y_test, y_pred, zero_division=0):.4f}")
print(f"Recall    = {recall_score(y_test, y_pred, zero_division=0):.4f}")
print(f"F1-score  = {f1_score(y_test, y_pred, zero_division=0):.4f}")

fig, ax = plt.subplots()
ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax)
ax.set_title("RBF SVM Confusion Matrix")
plt.show()
